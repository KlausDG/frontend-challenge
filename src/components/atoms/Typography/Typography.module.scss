@use '@/assets/css/_tokens.scss' as *;

.typography {
  margin: 0;
  font-family: $font-family-base;
}

.h2 {
  font-size: $font-size-subtitle;
  font-weight: $font-weight-regular;
  line-height: $line-height-subtitle;
}

.h4 {
  font-size: $font-size-h4;
  font-weight: $font-weight-bold;
  line-height: $line-height-h4;
}

.h5 {
  font-size: $font-size-h5;
  font-weight: $font-weight-semibold;
  line-height: $line-height-h5;
}

.h6 {
  font-size: $font-size-h6;
  font-weight: $font-weight-bold;
  line-height: $line-height-h6;
}

.pre {
  font-size: $font-size-body;
  font-weight: $font-weight-regular;
}

.caption {
  font-size: $font-size-caption;
  font-weight: $font-weight-regular;
  line-height: $line-height-caption;
}

.semibold {
  font-size: $font-size-body;
  font-weight: $font-weight-semibold;
}

.primary {
  color: $color-primary;
}

.secondary {
  color: $color-secondary;
}

.text {
  color: $color-neutral;
}

.dark {
  color: $color-neutral-dark;
}

.success {
  color: $color-success;
}
