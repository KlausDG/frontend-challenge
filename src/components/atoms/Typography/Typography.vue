<script setup lang="ts">
import styles from './Typography.module.scss'
import { computed } from 'vue'

type Variant = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'body' | 'caption' | 'semibold'
type Color = 'primary' | 'secondary' | 'dark' | 'text' | 'success'

interface Props {
  tag?: string
  variant?: Variant
  color?: Color
}

const props = withDefaults(defineProps<Props>(), {
  tag: 'p',
  color: 'text',
})

const tagToVariantMap: Record<string, Props['variant']> = {
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  h4: 'h4',
  h5: 'h5',
  h6: 'h6',
  p: 'caption',
  pre: 'body',
}

const computedVariant = computed(() => props.variant ?? tagToVariantMap[props.tag ?? 'p'] ?? 'body')
const variantClass = computed(() => styles[computedVariant.value])
const colorClass = computed(() => styles[props.color ?? 'text'])
</script>

<template>
  <component :is="tag" :class="['typography', variantClass, colorClass]">
    <slot />
  </component>
</template>

<style module lang="scss">
@use './Typography.module.scss';
</style>
