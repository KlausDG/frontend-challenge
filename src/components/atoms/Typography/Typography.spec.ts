import { mount } from '@vue/test-utils'
import { describe, it, expect, vi } from 'vitest'
import Typography from './Typography.vue'

// Mock CSS module to simplify class checks
vi.mock('./Typography.module.scss', () => ({
  default: {
    h1: 'h1',
    h2: 'h2',
    h3: 'h3',
    h4: 'h4',
    h5: 'h5',
    h6: 'h6',
    body: 'body',
    caption: 'caption',
    semibold: 'semibold',
    primary: 'primary',
    secondary: 'secondary',
    dark: 'dark',
    text: 'text',
    success: 'success',
  },
}))

describe('Typography.vue', () => {
  it('renders slot content correctly', () => {
    const wrapper = mount(Typography, {
      slots: {
        default: 'Hello Typography',
      },
    })
    expect(wrapper.text()).toBe('Hello Typography')
  })

  it('renders with default tag "p"', () => {
    const wrapper = mount(Typography)
    expect(wrapper.element.tagName.toLowerCase()).toBe('p')
  })

  it('renders with the tag prop if provided', () => {
    const wrapper = mount(Typography, {
      props: { tag: 'h3' },
    })
    expect(wrapper.element.tagName.toLowerCase()).toBe('h3')
  })

  it('applies default variant class based on tag mapping', () => {
    const wrapper = mount(Typography, { props: { tag: 'h1' } })
    expect(wrapper.classes()).toContain('h1')
  })

  it('applies variant prop over tag mapping', () => {
    const wrapper = mount(Typography, {
      props: { tag: 'h1', variant: 'semibold' },
    })
    expect(wrapper.classes()).toContain('semibold')
  })

  it('applies default color class "text"', () => {
    const wrapper = mount(Typography)
    expect(wrapper.classes()).toContain('text')
  })

  it('applies color prop class correctly', () => {
    const wrapper = mount(Typography, {
      props: { color: 'primary' },
    })
    expect(wrapper.classes()).toContain('primary')
  })

  it('matches the component snapshot', () => {
    const wrapper = mount(Typography, {
      props: { tag: 'h2', variant: 'h2', color: 'dark' },
      slots: { default: 'Snapshot Typography' },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
