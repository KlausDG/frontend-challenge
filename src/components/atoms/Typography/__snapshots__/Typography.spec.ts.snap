// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Typography.vue > matches the component snapshot 1`] = `"<h2 class="typography h2 dark">Snapshot Typography</h2>"`;
